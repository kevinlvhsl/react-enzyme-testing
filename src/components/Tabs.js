import React, { Component } from 'react'
import PropTypes from 'prop-types'
/**
 * 接受内容展示
 * 有一个默认选中项
 */
export class Tabs extends Component {
  static propTypes = {
    activeIndex: PropTypes.number,
    onTabChange: PropTypes.func.isRequired
  }
  constructor(props) {
    super(props)
    this.state = {
      // activeIndex: props.activeIndex || 0
    }
  }
  changeTab = (e, index) => {
    e.preventDefault()
    if (this.props.activeIndex === index) return

    // this.setState({ activeIndex: index })
    this.props.onTabChange(index)
  }
  render() {
    const children = this.props.children
    const { activeIndex } = this.props
    return (
      <ul className="nav nav-tabs nav-fill my-4">
        {
          React.Children.map(children, (child, index) => {
            const activeClassName = index === activeIndex ? 'nav-link active' : 'nav-link'
            return (<li className="nav-item">
              <a href="#"
                onClick={(e) => {this.changeTab(e, index)}}
                className={activeClassName}
                role="button"
              >
              {child}
              </a>
            </li>)
          })
        }
      </ul>
    )
  }
}

export const Tab = ({ children }) =>
  (<React.Fragment>{children}</React.Fragment>)