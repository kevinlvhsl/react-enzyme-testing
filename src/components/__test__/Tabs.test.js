import React from 'react'
import { shallow } from 'enzyme'
import { Tabs, Tab } from '../Tabs'

const props = {
  activeIndex: 0,
  onTabChange: jest.fn()
}
let wrapper

describe('test Tabs and Tab Componnet', () => {
  beforeEach(() => {
    wrapper = shallow(
      <Tabs {...props}>
        <Tab>1</Tab>
        <Tab>2</Tab>
      </Tabs>
    )
  })

  it('should render the component to match the snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })
  it('should render two Tab component, first one should be active', () => {
    // const tabs = wrapper.find(Tab)
    expect(wrapper.find(Tab).length).toEqual(2)
    expect(wrapper.find('.nav-link').length).toEqual(2)
    expect(wrapper.state().activeIndex).toEqual(0)
    expect(wrapper.find('.nav-link').first().hasClass('active')).toEqual(true)
  })

  it('click the 2nd Tab should change the active status and trigger the right function', () => {
    wrapper.find('.nav-link').last().simulate('click', { preventDefault: () => {} })
    expect(wrapper.state().activeIndex).toEqual(1)
    expect(wrapper.find('.nav-link').first().hasClass('active')).toEqual(false)
    expect(wrapper.find('.nav-link').last().hasClass('active')).toEqual(true)
    expect(props.onTabChange).toHaveBeenCalledWith(1)
  })
})