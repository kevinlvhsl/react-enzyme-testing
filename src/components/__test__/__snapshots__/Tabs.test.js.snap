// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Tabs and Tab Componnet should render the component to match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Tabs
    activeIndex={0}
    onTabChange={[MockFunction]}
  >
    <Tab>
      1
    </Tab>
    <Tab>
      2
    </Tab>
  </Tabs>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <li
          className="nav-item"
        >
          <a
            className="nav-link active"
            href="#"
            onClick={[Function]}
            role="button"
          >
            <Tab>
              1
            </Tab>
          </a>
        </li>,
        <li
          className="nav-item"
        >
          <a
            className="nav-link"
            href="#"
            onClick={[Function]}
            role="button"
          >
            <Tab>
              2
            </Tab>
          </a>
        </li>,
      ],
      "className": "nav nav-tabs nav-fill my-4",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": ".0",
        "nodeType": "host",
        "props": Object {
          "children": <a
            className="nav-link active"
            href="#"
            onClick={[Function]}
            role="button"
          >
            <Tab>
              1
            </Tab>
          </a>,
          "className": "nav-item",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <Tab>
              1
            </Tab>,
            "className": "nav-link active",
            "href": "#",
            "onClick": [Function],
            "role": "button",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": "1",
            },
            "ref": null,
            "rendered": "1",
            "type": [Function],
          },
          "type": "a",
        },
        "type": "li",
      },
      Object {
        "instance": null,
        "key": ".1",
        "nodeType": "host",
        "props": Object {
          "children": <a
            className="nav-link"
            href="#"
            onClick={[Function]}
            role="button"
          >
            <Tab>
              2
            </Tab>
          </a>,
          "className": "nav-item",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <Tab>
              2
            </Tab>,
            "className": "nav-link",
            "href": "#",
            "onClick": [Function],
            "role": "button",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": "2",
            },
            "ref": null,
            "rendered": "2",
            "type": [Function],
          },
          "type": "a",
        },
        "type": "li",
      },
    ],
    "type": "ul",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <li
            className="nav-item"
          >
            <a
              className="nav-link active"
              href="#"
              onClick={[Function]}
              role="button"
            >
              <Tab>
                1
              </Tab>
            </a>
          </li>,
          <li
            className="nav-item"
          >
            <a
              className="nav-link"
              href="#"
              onClick={[Function]}
              role="button"
            >
              <Tab>
                2
              </Tab>
            </a>
          </li>,
        ],
        "className": "nav nav-tabs nav-fill my-4",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": ".0",
          "nodeType": "host",
          "props": Object {
            "children": <a
              className="nav-link active"
              href="#"
              onClick={[Function]}
              role="button"
            >
              <Tab>
                1
              </Tab>
            </a>,
            "className": "nav-item",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <Tab>
                1
              </Tab>,
              "className": "nav-link active",
              "href": "#",
              "onClick": [Function],
              "role": "button",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": "1",
              },
              "ref": null,
              "rendered": "1",
              "type": [Function],
            },
            "type": "a",
          },
          "type": "li",
        },
        Object {
          "instance": null,
          "key": ".1",
          "nodeType": "host",
          "props": Object {
            "children": <a
              className="nav-link"
              href="#"
              onClick={[Function]}
              role="button"
            >
              <Tab>
                2
              </Tab>
            </a>,
            "className": "nav-item",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <Tab>
                2
              </Tab>,
              "className": "nav-link",
              "href": "#",
              "onClick": [Function],
              "role": "button",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": "2",
              },
              "ref": null,
              "rendered": "2",
              "type": [Function],
            },
            "type": "a",
          },
          "type": "li",
        },
      ],
      "type": "ul",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
